ringvax/__init__.py

```diff
@@ -47,52 +47,45 @@ def run(self) -> None:
         # queue is pairs (t_exposed, infector)
         # start with the index infection
         infection_queue: List[tuple[float, Optional[str]]] = [(0.0, None)]
-        termination = {"criterion": "", "last_complete_generation": -1}
-
-        while (
-            len(infection_queue) > 0
-            and len(self.query_people()) < self.params["max_infections"]
-        ):
-            # find the person who is infected next. this ensures that, if we stop the
-            # simulation after some number of infections, the first infections were done
-            # first
+
+        passed_max_generations = False
+
+        while True:
+            # Check if we need to stop the loop
+            n_infections = len(self.query_people())
+
+            if len(infection_queue) == 0 and not passed_max_generations:
+                self.termination = "extinct"
+                break
+            elif len(infection_queue) == 0 and passed_max_generations:
+                self.termination = "max_generations"
+                break
+            elif n_infections == self.params["max_infections"]:
+                self.termination = "max_infections"
+                break
+            elif n_infections > self.params["max_infections"]:
+                raise RuntimeError("Maximum number of infections exceeded")
```

Comment:
This strikes me as bad, and it's definitely not the logic we were using before.

Say we have a maximum of 100, and we're processing the 99th right now. This person produces a random number of offspring which could easily exceed 1, and that would make this error out rather than stop.

Reply:
To use "stop at `n` exactly" logic you need to operate at the level of adding to the queue.

Reply:
Put another way, this takes an expected occurrence and elevates it to a `RuntimeError`

Reply:
I don't think this is true. Each iteration of the `while True` loop adds (at most) a single infection to the simulation, so we should be able to stop at exactly the maximum number. Each iteration adds a potentially unlimited number of infections to the *queue*, but we pop people off the queue one by one.

Reply:
This sort of thing also makes me really prefer things that can be expressed in Markov terms, where you never have the future stored and can't get tripped up on the difference between the present and the future